File: 0083_Remove_Duplicates_from_Sorted_List/solution.java
package cn.huang.leetcode;

/*
    83. Remove Duplicates from Sorted List
    Given a sorted linked list, delete all duplicates such that each element appear only once.

    Example 1:

    Input: 1->1->2
    Output: 1->2

    Example 2:

    Input: 1->1->2->3->3
    Output: 1->2->3
 */
import cn.huang.leetcode.common.ListNode;

public class LC_0083_RemoveDuplicatesfromSortedList {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null)
            return head;

        ListNode slow = head;
        ListNode fast = slow.next;

        while(fast != null)
        {
            if(fast.val == slow.val)
            {
                fast = fast.next;
                continue;
            }
            slow.next = fast;
            slow = fast;
            fast = fast.next;
        }
        slow.next = fast;
        return head;
    }

    public ListNode deleteDuplicates_much_better(ListNode head) {
        if(head == null)
            return head;

        ListNode current = head;

        while(current != null && current.next != null)
        {
            if(current.val == current.next.val)
                current.next = current.next.next;
            else
                current = current.next;
        }

        return head;
    }
}

File: 0083_Remove_Duplicates_from_Sorted_List/solution.py
'''
83. Remove Duplicates from Sorted List
Given a sorted linked list, delete all duplicates such that each element appear only once.

For example,
Given 1->1->2, return 1->2.
Given 1->1->2->3->3, return 1->2->3.
'''


class Solution(object):
    def deleteDuplicates_my_not_prefect(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        prev = head
        curr = head.next
        while curr:
            if prev.val == curr.val:
                prev.next = curr.next
                curr.next = None
                curr = prev.next
            else:
                prev = prev.next
                curr = curr.next
        return head


def delete_duplicates_good(head):
    if head == None:
        return None

    current = head
    while current.next:
        next = current.next
        if next.val == current.val:
            current.next = current.next.next
        else:
            current = next
    return head
